File: rust/bron_kerbosch/src/bron_kerbosch1.rs
//! Naive Bron-Kerbosch algorithm

use graph::{connected_vertices, UndirectedGraph, VertexSetLike};
use reporter::{Clique, Reporter};

pub fn explore<VertexSet, Graph, Rprtr>(graph: &Graph, reporter: &mut Rprtr)
where
    VertexSet: VertexSetLike,
    Graph: UndirectedGraph<VertexSet = VertexSet>,
    Rprtr: Reporter,
{
    let candidates = connected_vertices(graph);
    if !candidates.is_empty() {
        visit(graph, reporter, candidates, VertexSet::new(), Clique::new());
    }
}

fn visit<VertexSet, Graph, Rprtr>(
    graph: &Graph,
    reporter: &mut Rprtr,
    mut candidates: VertexSet,
    mut excluded: VertexSet,
    clique: Clique,
) where
    VertexSet: VertexSetLike,
    Graph: UndirectedGraph<VertexSet = VertexSet>,
    Rprtr: Reporter,
{
    debug_assert!(candidates.all(|&v| graph.degree(v) > 0));
    debug_assert!(excluded.all(|&v| graph.degree(v) > 0));
    debug_assert!(candidates.is_disjoint(&excluded));

    if candidates.is_empty() {
        if excluded.is_empty() {
            reporter.record(clique);
        }
        return;
    }
    while let Some(v) = candidates.pop_arbitrary() {
        let neighbours = graph.neighbours(v);
        let neighbouring_candidates = neighbours.intersection_collect(&candidates);
        let neighbouring_excluded = neighbours.intersection_collect(&excluded);
        visit(
            graph,
            reporter,
            neighbouring_candidates,
            neighbouring_excluded,
            [clique.as_slice(), &[v]].concat(),
        );
        excluded.insert(v);
    }
}

File: go/src/bron_kerbosch/bron_kerbosch1.go
package bron_kerbosch

func bron_kerbosch1(graph *UndirectedGraph, reporter Reporter) {
	// Naive Bron-Kerbosch algorithm
	candidates := graph.connected_vertices()
	if !candidates.IsEmpty() {
		excluded := make(VertexSet, len(candidates))
		bron_kerbosch1_visit(
			graph,
			reporter,
			&candidates,
			&excluded,
			[]Vertex{})
	}
}

func bron_kerbosch1_visit(graph *UndirectedGraph, reporter Reporter, candidates *VertexSet,
	excluded *VertexSet, clique []Vertex) {
	for {
		v := candidates.PopArbitrary()
		neighbours := &graph.adjacencies[v]
		neighbouring_candidates := candidates.Intersection(neighbours)
		if !neighbouring_candidates.IsEmpty() {
			neighbouring_excluded := excluded.Intersection(neighbours)
			bron_kerbosch1_visit(graph, reporter,
				&neighbouring_candidates,
				&neighbouring_excluded,
				append(clique, v))
		} else {
			if excluded.IsDisjoint(neighbours) {
				reporter.Record(append(clique, v))
			}
			if candidates.IsEmpty() {
				break
			}
		}
		excluded.Add(v)
	}
}
